### ISSUE TITLE
No primal primal bound?

### ISSUE THREAD

**Issue by BYS543** [atoptima/Coluna.jl#545]
Hello,
I find there are warnings and errors when I try to set
```
coluna=optimizer_with_attributes(
Coluna.Optimizer,
"params"=>Coluna.Params(
solver=Coluna.Algorithm.TreeSearchAlgorithm()
),
"default_optimizer"=>Gurobi.Optimizer
)
```
and solve the model,which is as follows,
```
Coluna
Version 0.3.9 | 2021-05-15 | https://github.com/atoptima/Coluna.jl
┌ Warning: No initial primal bound and no cost for global artificial variables.
│ Cost of global artificial variables set to 100000.0
└ @ Coluna C:\Users\Lenovo\.julia\packages\Coluna\S4YuA\src\optimize.jl:20
┌ Warning: No initial primal bound and no cost for local artificial variables.
│ Cost of local artificial variables set to 10000.0
└ @ Coluna C:\Users\Lenovo\.julia\packages\Coluna\S4YuA\src\optimize.jl:33
ERROR: LoadError: Cannot apply run! for arguments Coluna.Algorithm.TreeSearchAlgorithm, Coluna.Algorithm.ModelData, Coluna.Algorithm.OptimizationInput{Coluna.MathProg.Formulation{Coluna.MathProg.Original},Coluna.MathProg.MinSense}.
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] run!(::Coluna.Algorithm.TreeSearchAlgorithm, ::Env, ::Coluna.Algorithm.ModelData, ::Coluna.Algorithm.OptimizationInput{Coluna.MathProg.Formulation{Coluna.MathProg.Original},Coluna.MathProg.MinSense}) at C:\Users\Lenovo\.julia\packages\Coluna\S4YuA\src\Algorithm\interface.jl:76
 [3] optimize!(::Coluna.MathProg.Formulation{Coluna.MathProg.Original}, ::Env, ::Coluna.ColunaBase.Bound{Coluna.MathProg.Primal,Coluna.MathProg.MinSense}, 
::Coluna.ColunaBase.Bound{Coluna.MathProg.Dual,Coluna.MathProg.MinSense}) at C:\Users\Lenovo\.julia\packages\Coluna\S4YuA\src\optimize.jl:138
 [4] macro expansion at C:\Users\Lenovo\.julia\packages\Coluna\S4YuA\src\optimize.jl:63 [inlined]
 [5] macro expansion at C:\Users\Lenovo\.julia\packages\TimerOutputs\PZq45\src\TimerOutput.jl:210 [inlined]
 [6] optimize!(::Env, ::Coluna.MathProg.Problem, ::Coluna.Annotations) at C:\Users\Lenovo\.julia\packages\Coluna\S4YuA\src\optimize.jl:62
 [7] optimize! at C:\Users\Lenovo\.julia\packages\Coluna\S4YuA\src\MOIwrapper.jl:98 [inlined]
 [8] optimize!(::MathOptInterface.Bridges.LazyBridgeOptimizer{Coluna.Optimizer}) at C:\Users\Lenovo\.julia\packages\MathOptInterface\YDdD3\src\Bridges\bridge_optimizer.jl:319
 [9] optimize!(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.GenericModel{Float64,MathOptInterface.Utilities.ModelFunctionConstraints{Float64}}}}) at C:\Users\Lenovo\.julia\packages\MathOptInterface\YDdD3\src\Utilities\cachingoptimizer.jl:252
 [10] optimize!(::Model, ::Nothing; bridge_constraints::Bool, ignore_optimize_hook::Bool, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at C:\Users\Lenovo\.julia\packages\JuMP\MIPb6\src\optimizer_interface.jl:185
 [11] optimize!(::Model) at C:\Users\Lenovo\.julia\packages\BlockDecomposition\05X3M\src\BlockDecomposition.jl:52
 [12] optimize!(::Model, ::Nothing; bridge_constraints::Bool, ignore_optimize_hook::Bool, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at C:\Users\Lenovo\.julia\packages\JuMP\MIPb6\src\optimizer_interface.jl:173
 [13] optimize! at C:\Users\Lenovo\.julia\packages\JuMP\MIPb6\src\optimizer_interface.jl:157 [inlined] (repeats 2 times)
```
Another research method
`solver=Coluna.Algorithm.SolveIpForm()`
there are no errors, but the same warnings as follows,
```
Version 0.3.9 | 2021-05-15 | https://github.com/atoptima/Coluna.jl
┌ Warning: No initial primal bound and no cost for global artificial variables.
│ Cost of global artificial variables set to 100000.0
└ @ Coluna C:\Users\Lenovo\.julia\packages\Coluna\S4YuA\src\optimize.jl:20
┌ Warning: No initial primal bound and no cost for local artificial variables.
│ Cost of local artificial variables set to 10000.0
└ @ Coluna C:\Users\Lenovo\.julia\packages\Coluna\S4YuA\src\optimize.jl:33
┌ Warning: Solver has no result to show.
└ @ Coluna.Algorithm C:\Users\Lenovo\.julia\packages\Coluna\S4YuA\src\Algorithm\basic\solveipform.jl:130
No primal solution found. Termination status is INFEASIBLE. 
 ──────────────────────────────────────────────────────────────────
                           Time                   Allocations      
                   ──────────────────────   ───────────────────────
 Tot / % measured:      3874s / 0.04%           6.22GiB / 1.66%

 Section   ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────
 Coluna         2    1.44s   100%   721ms    106MiB  100%   52.8MiB
 ──────────────────────────────────────────────────────────────────
[ Info: Terminated
[ Info: Primal bound: Inf
[ Info: Dual bound: -Inf
```
The model I test is not blocked first(all variables and constraints are in the master block) and if I don't use Coluna, the model is feasible. Why it is infeasible after I use Coluna to solve? By the way, what does these warnings and errors mean? 
Thanks in  advance!

**Comment by guimarqu** [comment]
Hi,

The warning at the beginning are ok. Maybe we should transform them into classic logs.
Coluna shows these information to the user because if the cost of artificial variables is too low (it depends on your problem), the algorithm may always find solutions with artifical variables and it may stuck or return infeasible.

If you don't decompose your problem you can't use `TreeSearchAlgorithm` because you can use this algorithm only after a Dantzig-Wolfe decomposition. You have to call `SolveIpForm` that will run the `default_optimizer` to optimize your original formulation : 

```julia
solver=Coluna.Algorithm.SolveIpForm()
```

I think we should improve the error message.

**Comment by BYS-NCEPU** [comment]
Thank you for your prompt reply. So if I want to use `SolveIpForm()`, I should ensure the cost of artificial variables are not too low? I am not familiar with the process of B&P algorithm, the cost of artificial variables mean what? If I use the block decomposition and use `TreeSearchAlgorithm`, will the cost of artificial variables disappear?

**Comment by guimarqu** [comment]
olveIpForm won't use the artificial variables because this algorithm is just a call to the optimizer. In that case, Coluna does nothing. It's just a shortcut when you want to try without decomposition.

Artificial variables are used by the branch and price. It's an additional variable that we add to each constraint to make sure that the master LP is not infeasible. It's also useful when we want to prove infeasibility.

You should not pay attention to the warnings.

**Comment by BYS-NCEPU** [comment]
Ok, thank you!

**Comment by BYS-NCEPU** [comment]
Hi, Guillaume
    There may be a problem for me of SolveIpForm, why it has no solution when I don't decompose the problem,just use the solver to solve my model, the code i use just the example 
```
nb_machines = 4
nb_jobs = 30
c = [12.7 22.5 8.9 20.8 13.6 12.4 24.8 19.1 11.5 17.4 24.7 6.8 21.7 14.3 10.5 15.2 14.3 12.6 9.2 20.8 11.7 17.3 9.2 20.3 11.4 6.2 13.8 10.0 20.9 20.6;  19.1 24.8 24.4 23.6 16.1 20.6 15.0 9.5 7.9 11.3 22.6 8.0 21.5 14.7 23.2 19.7 19.5 7.2 6.4 23.2 8.1 13.6 24.6 15.6 22.3 8.8 19.1 18.4 22.9 8.0;  18.6 14.1 22.7 9.9 24.2 24.5 20.8 12.9 17.7 11.9 18.7 10.1 9.1 8.9 7.7 16.6 8.3 15.9 24.3 18.6 21.1 7.5 16.8 20.9 8.9 15.2 15.7 12.7 20.8 10.4;  13.1 16.2 16.8 16.7 9.0 16.9 17.9 12.1 17.5 22.0 19.9 14.6 18.2 19.6 24.2 12.9 11.3 7.5 6.5 11.3 7.8 13.8 20.7 16.8 23.6 19.1 16.8 19.3 12.5 11.0]
w = [61 70 57 82 51 74 98 64 86 80 69 79 60 76 78 71 50 99 92 83 53 91 68 61 63 97 91 77 68 80; 50 57 61 83 81 79 63 99 82 59 83 91 59 99 91 75 66 100 69 60 87 98 78 62 90 89 67 87 65 100; 91 81 66 63 59 81 87 90 65 55 57 68 92 91 86 74 80 89 95 57 55 96 77 60 55 57 56 67 81 52;  62 79 73 60 75 66 68 99 69 60 56 100 67 68 54 66 50 56 70 56 72 62 85 70 100 57 96 69 65 50]
Q = [1020 1460 1530 1190]
using JuMP, BlockDecomposition, Coluna, Gurobi
coluna = optimizer_with_attributes(
    Coluna.Optimizer,
    "params" => Coluna.Params(
        solver = Coluna.Algorithm.SolveIpForm() 
    ),
    "default_optimizer" => Gurobi.Optimizer 
)

@axis(M, 1:nb_machines)
J = 1:nb_jobs

model = BlockModel(coluna)
@variable(model, x[m in M, j in J], Bin)
@constraint(model, cov[j in J], sum(x[m, j] for m in M) >= 1)
@constraint(model, knp[m in M], sum(w[m, j] * x[m, j] for j in J) <= Q[m])
@objective(model, Min, sum(c[m, j] * x[m, j] for m in M, j in J))

@dantzig_wolfe_decomposition(model, decomposition, M)

master = getmaster(decomposition)
subproblems = getsubproblems(decomposition)

specify!.(subproblems, lower_multiplicity = 0, upper_multiplicity = 1)

optimize!(model)
```
However, it shows as,
```
Coluna
Version 0.3.9 | 2021-05-15 | https://github.com/atoptima/Coluna.jl
No primal solution found. Termination status is INFEASIBLE.
 ──────────────────────────────────────────────────────────────────
                           Time                   Allocations
                   ──────────────────────   ───────────────────────
 Tot / % measured:      92.9s / 0.25%            375MiB / 5.50%

 Section   ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────
 Coluna         1    236ms   100%   236ms   20.6MiB  100%   20.6MiB
 ──────────────────────────────────────────────────────────────────
[ Info: Terminated
[ Info: Primal bound: Inf
[ Info: Dual bound: -Inf
```

**Comment by guimarqu** [comment]
Hi,

It looks like you decompose. You have to comment the following line :

```julia
@dantzig_wolfe_decomposition(model, decomposition, M)
```

**Comment by BYS-NCEPU** [comment]
Thank you, i just want to solve my model without decomposing, 
```
coluna = optimizer_with_attributes(
    Coluna.Optimizer,
    "params" => Coluna.Params(
        solver = Coluna.Algorithm.SolveIpForm() 
    ),
    "default_optimizer" => Gurobi.Optimizer
)
model = BlockModel(coluna)
@variable(model, ......)
@constraint(model, .......)
@objective(model, Min, ......)
optimize!(model)
```
However, it shows my model is infeasible, however, when i directly use
```
model = Model(Gurobi.Optimizer)
@variable(model, ......)
@constraint(model, .......)
@objective(model, Min, ......)
optimize!(model)
```
it can be solved. 
Moreover, if i use the example
```
coluna = optimizer_with_attributes(
    Coluna.Optimizer,
    "params" => Coluna.Params(
        solver = Coluna.Algorithm.SolveIpForm() # default BCP
    ),
    "default_optimizer" => Gurobi.Optimizer # GLPK for the master & the subproblems
)
M = 1:nb_machines
J = 1:nb_jobs
model = BlockModel(coluna)
@variable(model, x[m in M, j in J], Bin)
@constraint(model, cov[j in J], sum(x[m, j] for m in M) >= 1)
@constraint(model, knp[m in M], sum(w[m, j] * x[m, j] for j in J) <= Q[m])
@objective(model, Min, sum(c[m, j] * x[m, j] for m in M, j in J))
optimize!(model)

```
which shows
```
Found primal solution of 307.5000
 ──────────────────────────────────────────────────────────────────
                           Time                   Allocations
                   ──────────────────────   ───────────────────────
 Tot / % measured:     37897s / 0.01%           4.79GiB / 2.69%

 Section   ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────
 Coluna         1    1.91s   100%   1.91s    132MiB  100%    132MiB
 ──────────────────────────────────────────────────────────────────
[ Info: Terminated
[ Info: Primal bound: 307.50000000000006
[ Info: Dual bound: 307.50000000000006
```
it can be solved after 1.91s compared with 77.8s using decompose algorithm(TreeSearchAlgorithm), (1) why it becomes slow after using the decompose algorithm.

**Comment by guimarqu** [comment]
Hi,

The first issue should not happen. Do you use the Generalized Assignment Problem ? Do you have a minimal working example to reproduce the bug ?

The second one happens. I think it's quite difficult to compete with Gurobi on the Generalized Assignment Problem if the formulation is of little or medium size and if you use a simple branch-and-price (default TreeSearchAlgorithm).

To make TreeSearchAlgorithm more performant, you can strengthen your formulation with cut generation and
use an efficient algorithm to solve the subproblem (such as https://github.com/rafaelmartinelli/KnapsackLib.jl).
